File: client/java/src/main/java/io/be/mesh/prsim/KV.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package io.be.mesh.prsim;

import io.be.mesh.macro.MPI;
import io.be.mesh.macro.SPI;
import io.be.mesh.mpc.Types;
import io.be.mesh.struct.Entity;

import java.util.Optional;

/**
 * @author coyzeng@gmail.com
 */
@SPI("mesh")
public interface KV {

    // Get the value from kv store.
    @MPI("mesh.kv.get")
    Entity get(String key);

    // Put the value to kv store.
    @MPI("mesh.kv.put")
    void put(String key, Entity value);

    // Remove the kv store.
    @MPI("mesh.kv.remove")
    void remove(String key);

    /**
     * Get by codec.
     */
    default <T> T get(String key, Types<T> type) {
        return Optional.ofNullable(this.get(key)).orElseGet(Entity::empty).tryReadObject(type);
    }

    /**
     * Put by codec.
     */
    default void put(String key, Object value) {
        this.put(key, Entity.wrap(value));
    }
}

File: client/python/mesh/prsim/kv.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#

from abc import ABC, abstractmethod
from typing import Type, Any

from mesh.kinds import Entity
from mesh.macro import mpi, spi, T


@spi("mesh")
class KV(ABC):

    @abstractmethod
    @mpi("mesh.kv.get")
    def get(self, key: str) -> Entity:
        """
        Get the value from kv store.
        :param key:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.kv.put")
    def put(self, key: str, value: Entity):
        """
        Put the value to kv store.
        :param key:
        :param value:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.kv.remove")
    def remove(self):
        """
        Remove the kv store.
        :return:
        """
        pass

    def get_with_type(self, key: str, kind: Type[T]) -> T:
        """
        Get by codec.
        :param key:
        :param kind:
        :return:
        """
        entity = self.get(key)
        return entity.try_read_object(kind) if entity else None

    def put_object(self, key: str, value: Any):
        """
        Put by codec.
        :param key:
        :param value:
        :return:
        """
        self.put(key, Entity.wrap(value))
